File: Problem3/ruby/lib/largest_prime_factor.rb
class LargestPrimeFactor
  def largest_prime_factor n
    prime_factors(n).max
  end
  
  def prime_factors n
    factors, divisor = [], 1
    while n > 1 and divisor += 1
      factors << divisor and n /= divisor while n % divisor == 0
    end
    factors
  end
end

File: Problem3/haskell/largest_prime_factor.hs
largest_prime_factor :: Integer -> Integer
largest_prime_factor num = maximum(prime_factors num)

prime_factors :: Integer -> [Integer]
prime_factors num
  | num == 1 = []
  | otherwise = p : prime_factors (num `div` p) where p = lowest_divisor num

lowest_divisor :: Integer -> Integer
lowest_divisor num = lowest_divisor_from 2 num

lowest_divisor_from :: Integer -> Integer -> Integer
lowest_divisor_from threshold num 
  | divides threshold num = threshold
  | threshold ^ 2 > num = num
  | otherwise = lowest_divisor_from (threshold+1) num

divides :: Integer -> Integer -> Bool
divides divisor num = num `rem` divisor == 0
